### PR TITLE
Fix cache race condition in v3

### PR DESCRIPTION
**Pull request by abhishek58g**
pull request Issue #308

### CODE REVIEW

File: src/ADAL.PCL.Android/ADAL.PCL.Android.csproj [hunk 1]
```diff
@@ -15,7 +15,7 @@
     <AndroidResgenFile>Resources\Resource.Designer.cs</AndroidResgenFile>
     <GenerateSerializationAssemblies>Off</GenerateSerializationAssemblies>
     <AndroidUseLatestPlatformSdk>True</AndroidUseLatestPlatformSdk>
-    <TargetFrameworkVersion>v5.1</TargetFrameworkVersion>
+    <TargetFrameworkVersion>v6.0</TargetFrameworkVersion>
```
Comment (kpanwar) [inline src/ADAL.PCL.Android/ADAL.PCL.Android.csproj]: why did this get updated?

File: src/ADAL.PCL/TokenCache.cs [hunk 2]
```diff
@@ -46,7 +46,9 @@ public class TokenCache
         // We do not want to return near expiry tokens, this is why we use this hard coded setting to refresh tokens which are close to expiration.
         private const int ExpirationMarginInMinutes = 5;
 
-        private volatile bool hasStateChanged; 
+        private volatile bool hasStateChanged;
+
+        public Object cacheLock = new Object();
```
Comment (kpanwar) [inline src/ADAL.PCL/TokenCache.cs]: why is this public?

File: tests/Test.ADAL.Common/TokenCacheTests.cs [hunk 3]
```diff
@@ -689,5 +689,32 @@ public static void TokenCacheBackCompatTest(byte[] oldcache)
                 Verify.IsNull(value.UserAssertionHash);
             }
         }
+        public static void ParallelStorePositiveTest(byte[] oldcache)
+        {
+            for (int i = 0; i < 100; i++)
```
Comment (kpanwar) [inline tests/Test.ADAL.Common/TokenCacheTests.cs]: no need to run this 100 times.

File: src/ADAL.PCL/TokenCache.cs (hunk 2 above)
Comment (abhishek58g) [inline src/ADAL.PCL/TokenCache.cs]: Made Private

File: tests/Test.ADAL.Common/TokenCacheTests.cs (hunk 3 above)
Comment (abhishek58g) [inline tests/Test.ADAL.Common/TokenCacheTests.cs]: Removed the loop

**Comment by kpanwar** [comment]
you should put lock around all the methods. This will reduce bugs in long run.

File: src/ADAL.PCL.Android/ADAL.PCL.Android.csproj (hunk 1 above)
Comment (abhishek58g) [inline src/ADAL.PCL.Android/ADAL.PCL.Android.csproj]: Reverted back the change

**Comment by abhishek58g** [comment]
Added locks for all the methods except private ones.

**Comment by kpanwar** [comment]
I think you should consider putting a lock around all.

**Comment by abhishek58g** [comment]
Added locks for all the methods

**Pull request merged**